src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -20,52 +22,80 @@ class ManagedValueManager : JniRuntime.JniValueManager
 {
 	const DynamicallyAccessedMemberTypes Constructors = DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors;
 
-	Dictionary<int, List<IJavaPeerable>>?   RegisteredInstances = new Dictionary<int, List<IJavaPeerable>>();
+	readonly Dictionary<int, List<ReferenceTrackingHandle>> RegisteredInstances = new ();
+	readonly ConcurrentQueue<IntPtr> CollectedContexts = new ();
+
+	bool disposed;
+
+	static readonly SemaphoreSlim bridgeProcessingSemaphore = new(1, 1);
 
 	static Lazy<ManagedValueManager> s_instance = new (() => new ManagedValueManager ());
 	public static ManagedValueManager GetOrCreateInstance () => s_instance.Value;
 
-	ManagedValueManager ()
+	unsafe ManagedValueManager ()
 	{
+		// There can only be one instance of ManagedValueManager because we can call JavaMarshal.Initialize only once.
+		var mark_cross_references_ftn = RuntimeNativeMethods.clr_initialize_gc_bridge (&BridgeProcessingStarted, &BridgeProcessingFinished);
+		JavaMarshal.Initialize (mark_cross_references_ftn);
 	}
 
-	public override void WaitForGCBridgeProcessing ()
+	protected override void Dispose (bool disposing)
 	{
+		disposed = true;
+		base.Dispose (disposing);
 	}
 
-	public override void CollectPeers ()
+	void ThrowIfDisposed ()
 	{
-		if (RegisteredInstances == null)
+		if (disposed)
 			throw new ObjectDisposedException (nameof (ManagedValueManager));
+	}
 
-		var peers = new List<IJavaPeerable> ();
+	public override void WaitForGCBridgeProcessing()
+	{
+		bridgeProcessingSemaphore.Wait();
+		bridgeProcessingSemaphore.Release();
+	}
 
-		lock (RegisteredInstances) {
-			foreach (var ps in RegisteredInstances.Values) {
-				foreach (var p in ps) {
-					peers.Add (p);
-				}
+	public unsafe override void CollectPeers ()
+	{
+		ThrowIfDisposed ();
+
+		while (CollectedContexts.TryDequeue (out IntPtr contextPtr)) {
+			HandleContext* context = (HandleContext*)contextPtr;
```

Comment:
Should we validate `contextPtr`? Is it possible for it to be `IntPtr.Zero`?

Reply:
Nah, no need to add an assert - but can you add a comment describing why there can't be `IntPtr.Zero` here?